packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -0,0 +1,237 @@
+import xxhash from 'xxhash-wasm';
+import { extname } from 'node:path';
+import { getBuildOutputDir } from '../../core/build/util.js';
+import { mkdirSync, writeFileSync } from 'node:fs';
+import { generateFontFace } from './utils.js';
+import {
+	DEFAULTS,
+	VIRTUAL_MODULE_ID,
+	RESOLVED_VIRTUAL_MODULE_ID,
+	URL_PREFIX,
+	CACHE_DIR,
+} from './constants.js';
+import { removeTrailingForwardSlash } from '@astrojs/internal-helpers/path';
+import type { Logger } from '../../core/logger/core.js';
+import { createCache, createStorage, type Cache } from './cache.js';
+
+interface Options {
+	settings: AstroSettings;
+	sync: boolean;
+	logger: Logger;
+}
+
+type PreloadData = Array<{ url: string; type: string }>;
+
+const createLogManager = (logger: Logger) => {
+	const done = new Set<string>();
+	const items = new Set<string>();
+	let id: NodeJS.Timeout | null = null;
+
+	return {
+		add: (value: string) => {
+			if (done.has(value)) {
+				return;
+			}
+
+			if (items.size === 0 && id === null) {
+				logger.info('assets', 'Downloading fonts...');
+			}
+			items.add(value);
+			if (id) {
+				clearTimeout(id);
+				id = null;
+			}
+		},
+		remove: (value: string, cached: boolean) => {
+			if (done.has(value)) {
+				return;
+			}
+
+			items.delete(value);
+			done.add(value);
+			if (id) {
+				clearTimeout(id);
+				id = null;
+			}
+			id = setTimeout(() => {
+				let msg = 'Done';
+				if (cached) {
+					msg += ' (loaded from cache)';
+				}
+				logger.info('assets', msg);
+			}, 50);
+		},
+	};
+};
+
+export function fonts({ settings, sync, logger }: Options): Plugin {
+	if (!settings.config.experimental.fonts) {
+		// this is required because the virtual module does not exist
+		// when fonts are not enabled, and that prevents rollup from building
+		// TODO: remove once fonts are stabilized
+		return {
+			name: 'astro:fonts:fallback',
+			config() {
+				return {
+					build: {
+						rollupOptions: {
+							external: [VIRTUAL_MODULE_ID],
+						},
+					},
+				};
+			},
+		};
+	}
+
+	const providers: Array<FontProvider<string>> = settings.config.experimental.fonts.providers ?? [];
+	const families: Array<FontFamily<string>> = settings.config.experimental.fonts.families;
+
+	const baseUrl = removeTrailingForwardSlash(settings.config.base) + URL_PREFIX;
+
+	let resolvedMap: Map<string, { preloadData: PreloadData; css: string }> | null = null;
+	/** Key is `${hash}.${ext}`, value is a URL */
+	let collected: Map<string, string> | null = null;
+	let isBuild: boolean;
+	let cache: Cache['cache'] | null = null;
+
+	return {
+		name: 'astro:fonts',
+		config(_, { command }) {
+			isBuild = command === 'build';
+		},
+		async buildStart() {
+			const { h64ToString } = await xxhash();
+
+			const resolved = await resolveProviders({
+				settings,
+				providers,
+			});
+
+			const storage = createStorage({
+				// In build, first check the build cache then fallback to dev cache
+				// In dev, only check dev cache
```

Comment:
I am not sure this should be the expected behaviour. Usually, builds are done in CI environments, where the dev cache doesn't exist. If we use the dev cache in local machines, we might risk to create artefacts that don't match CI builds.